### PR TITLE
LUCENE-10504: KnnGraphTester to use KnnVectorQuery

### PR DESCRIPTION
**Pull request by msokolov**
This really has two changes:

1. it switches the vector searches it runs to use the Query impl, as the description says
2. it becomes a bit more clever about managing its cache of "exact" NN that are used for recall comparisons. Previously, if you changed the source data files it would still potentially re-use the cached NN file. Now it stores a hash of the file name and looks at the modification times to see if it should regenerate the NN file

### CODE REVIEW

File: lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java [hunk 1]
```diff
@@ -362,18 +367,19 @@ private void testSearch(Path indexPath, Path queryPath, Path outputPath, int[][]
       long cpuTimeStartNs;
       try (Directory dir = FSDirectory.open(indexPath);
           DirectoryReader reader = DirectoryReader.open(dir)) {
+        IndexSearcher searcher = new IndexSearcher(reader);
         numDocs = reader.maxDoc();
         for (int i = 0; i < warmCount; i++) {
           // warm up
           targets.get(target);
-          results[i] = doKnnSearch(reader, KNN_FIELD, target, topK, fanout);
+          doKnnSearch(reader, KNN_FIELD, target, topK, fanout);
```
Comment (jtibshirani) [inline lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java:375]: Would it be fine to use `doKnnVectorQuery` here so we could delete `doKnnSearch`?

File: lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java [hunk 2]
```diff
@@ -349,8 +353,9 @@ private void testSearch(Path indexPath, Path queryPath, Path outputPath, int[][]
     TopDocs[] results = new TopDocs[numIters];
     long elapsed, totalCpuTime, totalVisited = 0;
     try (FileChannel q = FileChannel.open(queryPath)) {
+      int bufferSize = Math.max(numIters, warmCount) * dim * Float.BYTES;
```
Comment (jtibshirani) [inline lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java:356]: Maybe we could just assert warmCount < numIters, seems unusual to warm up with queries that you don't use in the benchmark?

File: lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java (hunk 1 above)
Comment (msokolov) [reply lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java:375]: Yes, that makes sense, I don't see why not

File: lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java (hunk 2 above)
Comment (msokolov) [reply lucene/core/src/test/org/apache/lucene/util/hnsw/KnnGraphTester.java:356]: Yeah, I think we can simply replace warmCount with numIters

**Comment by jtibshirani** [comment]
@msokolov it'd be great to get this in when you have the time!

**Comment by msokolov** [comment]
> @msokolov it'd be great to get this in when you have the time!

whoa sorry I lost track. I will address the comments and push soon. I have other changes I want to make, relating to testing pre-filtering. Some folks here have been testing and finding very interesting results; I think we are going to want to have a good test harness for different filter selectivity.

**Comment by jtibshirani** [comment]
Thanks @msokolov ! So I'm not waiting in suspense too much, what sorts of interesting results have you found related to prefiltering?

**Comment by msokolov** [comment]
I'm trying not to steal the thunder of the folks who are actually working on this, but at a high level: we were seeing prefiltering being more expensive than postfiltering (over collecting and then filtering) for the same "yield" of top K, but were able to recover the cost, and flip the balance the other way by using the non-matching nodes to traverse the graph up until some threshold. Basically, we play with how we update the lower bound, allowing it to increase based on non-matching (the filter) nodes, until we get close to where we want to be. I guess the intuition is that nodes that are never going to be included in the top K are still useful for traversing the graph.

**Comment by jtibshirani** [comment]
Super interesting, looking forward to hearing more! I do hope we can stick with a prefiltering-like approach (and just improve its performance), since it feels easier to work with for users. If you request `k` documents, you always get `k` back -- there's no guessing about how many candidates you need to retrieve as in post-filtering.

This doesn't sound like what you're talking about, but I did notice that prefiltering can be expensive when the filter matches a lot of documents. Unlike in postfiltering, the filter is not allowed skip any of the docs and you need to convert all the matches into a bit set, which is not cheap.

**Pull request merged**